lib/charms/kubernetes_charm_libraries/v0/kubernetes_volumes_patch.py

```diff
@@ -0,0 +1,618 @@
+
+    @staticmethod
+    def _volume_is_hugepages(volume: Volume) -> bool:
+        """Returns whether the specified volume is HugePages."""
+        return volume.name.startswith("hugepages")
+
+    @staticmethod
+    def _limit_or_resource_is_hugepages(key: str) -> bool:
+        """Returns whether the specified limit or request regards HugePages."""
+        return key.startswith("hugepages")
+
+    def _configure_volumes(self):
+        """Configure HugePages in the StatefulSet and Pod.
+
+        1. Goes through the list of current volumeMounts for the specified container
+        - If list of requested hugepages is empty:
+          - If hugepages is set as volumeMount, remove it.
+          - Keep all other volumeMounts
+        - If list of requested hugepages is not empty:
+          - If hugepages is set as volumeMount, keep it.
+          - Else, add volumeMount.
+        2. Goes through the list of current volumes for the specified StatefulSet
+        - If list of requested hugepages is empty:
+          - If hugepages is set as volume, remove it.
+          - Keep all other volumes
+        - If list of requested hugepages is not empty:
+          - If hugepages is set as volume, keep it.
+          - Else, add volume.
+        3. Goes through the list of current resource requirements for the specified container
+        - If list of requested hugepages is empty:
+          - If hugepages is set in resource requirements, remove them.
+          - Keep all other resource requirements
+        - If list of requested hugepages is not empty:
+          - If hugepages is set in resource requirements, keep resource requirements.
+          - Else, add resource requirements.
+        """
+        # handle container volumeMounts
+        additional_volumemounts = self._generate_volumemounts_from_requested_hugepage()
+        current_volumemounts = self.kubernetes.list_volumemounts(
+            statefulset_name=self.model.app.name, container_name=self.container_name
+        )
+        for current_volumemount in current_volumemounts:
+            if not self._volumemount_is_hugepages(current_volumemount):
+                additional_volumemounts.append(current_volumemount)
+
+        # handle statefulset volumes
+        additional_volumes = self._generate_volumes_from_requested_hugepage()
+        current_volumes = self.kubernetes.list_volumes(
+            statefulset_name=self.model.app.name,
+        )
+        for current_volume in current_volumes:
+            if not self._volume_is_hugepages(current_volume):
+                additional_volumes.append(current_volume)
+
+        # handle resource requirements
+        additional_resources = self._generate_resource_requirements_from_requested_hugepage()
+        current_resources = self.kubernetes.list_container_resources(
+            statefulset_name=self.model.app.name, container_name=self.container_name
+        )
+
+        if self.hugepages_volumes_func():
```

Comment:
To me it looks like a good candidate for refactoring. It's really complex. Maybe it could be broken down into smaller functions?